Abstract keyboard shorcuts for heading to paragraph transform and vice-versa

afercia opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->
Part of https://github.com/WordPress/gutenberg/issues/60567

## What?
<!-- In a few words, what is the PR actually doing? -->
As a preparatory work for https://github.com/WordPress/gutenberg/issues/60567, this pull requests seeks to clean up and avoid code duplication for the keyboard shortcuts that are responsible for transforming heading levels and paragraph to heading and vice-versa.

Currently, the code responsible for this feature is duplicated in 4 different locations in the codebase: post editor, sit editor, widgets editor and customize widgets.

I'd appreciate some feedback about the 'architectural' choice of _where_ to move this code. See the 'How' below.
Cc @youknowriad @WordPress/gutenberg-core  when you have a chance 🙏🏽 

Additionally, this PR adds a new keyboard shortcut alias `7` to transform a Heading to Paragraph, for consistency with the Classic Editor where this shortcut has always used the number 7.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
Code duplication brings in maintenance cost and potential bugs and inconsistencies. Havin a centralized, unique, code responsible for a feature allow to easily change it, extend it, improve it, test it.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
- In a first moment I thought to abstract the related code to a new `BlockKeyboardShortcuts` component and add that to the stack. This was maybe cleaner but it added a new component.
- Then I discovered the other keyboard shortcuts for actions related to blocks (duplicate, move, etc.) are all already registered in the block-editor `KeyboardShortcuts` component and their 'callbacks' use a custom implementation in the `BlockTools` component. In the last commit I moved everything to follow this pattern. However, I'm not sure this existing implementation is ideal, as it mixes a component that does many other things (maybe too many) with the keyboard shortcuts.

Any suggestion on which of the two approaches would be better is welcome.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->
- Run `npm install` in your terminal before building, as the 'block-library' package has now a new dependency.
- Go to the Post Editor and select a Heading block.
- Check the following keyboard shortcuts work as expected (assuming macOS):
  - Control + Option + 1-6 change the heading level ot transform a Paragraph to a Heading with that level.
  - Control + Option + 0  and Control + Option + 7 change a Heading to a Paragraph.
- Go to the Site Editor and select a Heading block.
- Repeat the test.
- Set Twenty Twenty-One as active theme so that thre Widgets editor adn the Customizer are available.
- Go to Appearance > Widgets.
- Add a heading block to a widgets area and save.
- Select the heading block and repeat the test.
- Go to Customize > Widgets.
- Select the heading block and repeat the test.
- In all 4 cases, open the Keyboard shortcuts modal dialog by pressing Control + Option + H
- Observe the keyboard shortcut `Convert the current heading to a paragraph.` now displays also the new alias with thte key `7`.

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

afercia committed:
Abstract keyboard shorcuts for heading to paragraph transform and vice-versa.

afercia committed:
Add alias to transform heading to paragraph for consistency with classic editor.

youknowriad commented:
Thanks for opening the PR, indeed it seems that there's some duplication here. In some cases duplication is better than abstraction if we can't think of the right abstraction at a given time.

That said, for this one, It seems that the best place for the shared code for these is in the "block-library" package IMO because this is block specific code (some block-editors might not have paragraph or heading blocks or might not use the block library at all)

afercia commented:
@youknowriad thanks. So basically, follow the approach of [the first commit](https://github.com/WordPress/gutenberg/pull/60606/commits/c6eea9c5d8a56a636e53561c0b9eb8c9a7c7166d) with a new `<BlockKeyboardShortcuts />` component bt moved to the block-library package?

youknowriad commented:
@afercia yes, that would be a good first approach. The downside would be the need to "import" this component in each implementation. I kind of wish that registering the blocks should be enough for these shortcuts to work. Don't we have transform "types" that handle these for us already (cc @ellatrix) ?  If it's not the case, maybe fine to consider a separate issue for that.

ellatrix commented:
Yes, Ideally we reuse the block transforms API for this. Maybe add a shortcut prop to those objects?

afercia commented:
> Don't we have transform "types" that handle these ...

in https://github.com/WordPress/gutenberg/issues/60567 I would like to extend these keyboard shortctus to work also with other blocks e.g. the Site Title. These blocks don't use a 'transform'. Rather, the HTML tag they are rendered with is controlled by the block attribute `level`.

afercia committed:
Move code to block-library.

afercia commented:
Last commits moves shared code to the 'block-library' package. I had to add `@wordpress/keyboard-shortcuts` as dependency to this package. WIll update the test instructions.

youknowriad commented on packages/block-library/src/index.js:
```diff
@@ -330,3 +330,5 @@ export const __experimentalRegisterExperimentalCoreBlocks = process.env
 	  }
 	: undefined;
+
+export { default as BlockKeyboardShortcuts } from './block-keyboard-shortcuts';
```
Can we make it a private component for now. (until we make these transforms or something else)

afercia replied:
> Can we make it a private component for now. (until we make these transforms or something else)

@youknowriad thanks. Do you mean using lock/unlock? See last commit.

youknowriad replied:
yes, that was correct 👍

afercia committed:
Make BlockKeyboardShortcuts private.

youknowriad commented on packages/block-library/src/block-keyboard-shortcuts/index.js:
```diff
@@ -0,0 +1,96 @@
+import { createBlock } from '@wordpress/blocks';
+import { store as blockEditorStore } from '@wordpress/block-editor';
+
+function BlockKeyboardShortcuts() {
```
In this component registration and usage is merged. They were in different components before. The reason is that in some contexts, we want to register the shortcuts (show them in the keyboard shortcuts modal) but "disable" them in some contexts (view mode in the site editor for instance). 

I'm not sure whether this creates issues for this PR, it needs to be tested.

youknowriad replied:
It seems the `Block` prefix is redundant here as well, we're already in the block library package.

afercia replied:
I can rename it but then I'd have to import it 'as' a different name you know, because other components are already named KeyboardShortcuts.

youknowriad replied:
yeah, we do that already for other things. I don't care much though, it's a private API anyway.

afercia replied:
> In this component registration and usage is merged. They were in different components before. 

I think they were separated only in edit-site, where the `use` was [here](https://github.com/WordPress/gutenberg/blob/cca1f55f048f027df49f0c08e757e73f73a54193/packages/edit-site/src/components/keyboard-shortcuts/edit-mode.js#L71-L83)? Not sure that makes a difference, because they are within [a check for isEditMode](https://github.com/WordPress/gutenberg/blob/678938fc22d4f67c2890e73623adb7457125351b/packages/edit-site/src/components/editor/index.js#L283-L290)?

youknowriad replied:
yes, for this PR it probably doesn't change much.

youknowriad reviewed (commented):
This is looking good to me overall, I'd appreciate some testing for other folks.

afercia commented:
This PR now has merge conflicts because of relevant code reorganization from https://github.com/WordPress/gutenberg/pull/60866. Honestly, annoying.

youknowriad reviewed (approved):
Sorry for the delay approving.

I'm happy to help with the conflicts if needed. They should be smallish.

afercia committed:
Adjust after rebase.

afercia commented:
Rebased on top of latest trunk, hopefully tests will pass.
I removed `packages/edit-site/src/components/keyboard-shortcuts/edit-mode.js` as after https://github.com/WordPress/gutenberg/pull/60856 and after this PR it doesn't contain any feature.
